cassiopeia/solver/ILPSolver.py

```diff
@@ -0,0 +1,557 @@
+
+        logging.info(
+            f"(Process {pid}) Steiner tree solving tool {days} days, "
+            f"{hours} hours, {minutes} minutes, and {seconds} seconds."
+        )
+        if model.status != gurobipy.GRB.status.OPTIMAL:
+            logging.info(
+                f"(Process {pid}) Warning: Steiner tree solving did "
+                "not result in an optimal model."
+            )
+
+        return solutions
+
+    def post_process_steiner_solution(
+        self,
+        solution: nx.DiGraph,
+        root: List[int],
+        targets: List[List[int]],
+        pid: int,
+    ) -> nx.DiGraph:
+        """Post-processes the returned graph from Gurobi.
+
+        This procedure post-processes the proposed Steiner Tree from Gurobi
+        by enforcing that no self-loops occur and that every node at most one
+        parent.
+
+        Args:
+            solution: The Gurobi solution
+            root: The root node
+            targets: A list of targets
+            pid: Process id
+
+        Returns:
+            A cleaned up networkx solution 
+        """
+
+        processed_solution = solution.copy()
+        for edge in nx.selfloop_edges(processed_solution):
+            processed_solution.remove_edge(edge[0], edge[1])
+
+        non_tree_nodes = [
+            n
+            for n in processed_solution.nodes()
+            if processed_solution.in_degree(n) > 1
+        ]
+        for node in non_tree_nodes:
+            parents = processed_solution.predecessors(node)
+            parents = sorted(
+                parents,
+                key=lambda k: processed_solution[k][node]["weight"],
+                reverse=True,
+            )
+
+            if len(parents) == 2 and (
+                parents[1] in nx.ancestors(processed_solution, parents[0])
+                or (parents[0] in nx.ancestors(processed_solution, parents[1]))
+            ):
+                if parents[1] in nx.ancestors(processed_solution, parents[0]):
+                    processed_solution.remove_edge(parents[1], node)
+                else:
+                    processed_solution.remove_edge(parents[0], node)
+
+            else:
+                for parent in parents[1:]:
+                    processed_solution.remove_edge(parent, node)
+
+        # remove spurious roots
+        spurious_roots = [
+            n
+            for n in processed_solution
+            if processed_solution.in_degree(n) == 0
+        ]
+        while len(spurious_roots) > 1:
+            for r in spurious_roots:
+                if r != root:
+                    processed_solution.remove_node(r)
+            spurious_roots = [
+                n
+                for n in processed_solution
+                if processed_solution.in_degree(n) == 0
+            ]
+
+        processed_solution = self.append_sample_names(processed_solution)
+
+        return processed_solution
+
+    def append_sample_names(self, solution: nx.DiGraph) -> nx.DiGraph:
```

Comment:
So if I understand it, this function appends the samples as explicit nodes into the graph? So this will append all duplicate samples and also append explicit sample nodes such that no sample is an internal node in the final tree?

Reply:
yes, the purpose of this function is to append sample names onto their appropriate character states in the tree was just built. 

To reduce the complexity, we build the tree over unique character states. So we need to append sample names appropriately to the tree after we've built it for interpretability.

Reply:
Ah I realized I hadn't documented this function - I'll add documentation.

Reply:
Thanks Matt!